tezos/context/src/kv_store/persistent.rs

```diff
@@ -0,0 +1,532 @@
+
+        self.hashes.get_hash(hash_id)
+    }
+
+    fn get_vacant_object_hash(&mut self) -> Result<VacantObjectHash, DBError> {
+        self.hashes.get_vacant_object_hash()
+    }
+
+    fn memory_usage(&self) -> RepositoryMemoryUsage {
+        RepositoryMemoryUsage::default()
+    }
+
+    fn get_shape(&self, shape_id: DirectoryShapeId) -> Result<ShapeStrings, DBError> {
+        self.shapes
+            .get_shape(shape_id)
+            .map(ShapeStrings::SliceIds)
+            .map_err(Into::into)
+    }
+
+    fn make_shape(
+        &mut self,
+        dir: &[(StringId, DirEntryId)],
+    ) -> Result<Option<DirectoryShapeId>, DBError> {
+        self.shapes.make_shape(dir).map_err(Into::into)
+    }
+
+    fn get_str(&self, string_id: StringId) -> Option<&str> {
+        self.string_interner.get_str(string_id).ok()
+    }
+
+    fn synchronize_strings_from(&mut self, string_interner: &StringInterner) {
+        self.string_interner.extend_from(string_interner);
+    }
+
+    fn synchronize_strings_on_reload(&self, string_interner: &mut StringInterner) {
+        string_interner.clone_after_reload(&self.string_interner);
+    }
+
+    fn get_object(
+        &self,
+        object_ref: ObjectReference,
+        storage: &mut Storage,
+        strings: &mut StringInterner,
+    ) -> Result<Object, DBError> {
+        self.get_object_bytes(object_ref, &mut storage.data)?;
+
+        let object_bytes = std::mem::take(&mut storage.data);
+        let result = persistent::deserialize_object(
+            &object_bytes,
+            object_ref.offset(),
+            storage,
+            strings,
+            self,
+        );
+        storage.data = object_bytes;
+
+        result.map_err(Into::into)
+    }
+
+    fn get_object_bytes<'a>(
+        &self,
+        object_ref: ObjectReference,
+        buffer: &'a mut Vec<u8>,
+    ) -> Result<&'a [u8], DBError> {
+        self.get_object_bytes(object_ref, buffer)
+            .map_err(Into::into)
+    }
+
+    fn commit(
+        &mut self,
+        working_tree: &WorkingTree,
+        parent_commit_ref: Option<ObjectReference>,
+        author: String,
+        message: String,
+        date: u64,
+    ) -> Result<(ContextHash, Box<SerializeStats>), DBError> {
+        let offset = self.data_file.offset();
+
+        let PostCommitData {
+            commit_ref,
+            serialize_stats,
+            output,
+            ..
+        } = working_tree
+            .prepare_commit(
+                date,
+                author,
+                message,
+                parent_commit_ref,
+                self,
+                Some(persistent::serialize_object),
+                Some(offset),
+            )
+            .map_err(Box::new)?;
+
+        let commit_hash = get_commit_hash(commit_ref, self).map_err(Box::new)?;
+
+        self.commit_to_disk(&output)
+            .map_err(|err| DBError::CommitToDiskError { err })?;
+
+        self.put_context_hash(commit_ref)?;
```

Comment:
@sebastiencs  this here is happening after all the previous calls to `fsync()` right? cannot we lose that data then?

Reply:
Ah right, `put_context_hash` must be called before `commit_to_disk`